Sources/Tools/SoundDecompress/main.cpp

```diff
@@ -0,0 +1,258 @@
+//  Page iterator that collects all the plSoundBuffers in all of our pages
+
+class plSoundBufferCollector : public plRegistryPageIterator, public plKeyCollector
+{
+public:
+    plSoundBufferCollector(std::set<plKey>& keyArray)
+                : plKeyCollector(keyArray) {}
+
+    bool EatPage(plRegistryPageNode* page)
+    {
+        if (page->IsValid())
+        {
+            page->LoadKeys();
+            return page->IterateKeys(this, plSoundBuffer::Index());
+        }
+        else
+        {
+            fprintf(stderr, "INVALID PAGE: %s\n", page->GetPagePath().AsString().c_str());
+            return true;
+        }
+    }
+};
+
+
+void PrintHelp()
+{
+    printf("Plasma Sound Decompressor\n");
+    printf("-------------------------\n");
+    printf("-s, --silent        Run silently, no output\n");
+    printf("-v, --verbose       Print each filename when decompressing\n");
+    printf("-f, --force         Force decompressing existing files\n");
+}
+
+bool CollectSounds(plResManager* rm, SoundSet& sfxArray)
+{
+    std::set<plKey> soundKeys;
+
+    plSoundBufferCollector soundCollector(soundKeys);
+    rm->IterateAllPages(&soundCollector);
+
+    for (auto it = soundKeys.begin(); it != soundKeys.end(); ++it)
+    {
+        plSoundBuffer* buffer = plSoundBuffer::ConvertNoRef((*it)->VerifyLoaded());
+        if (buffer)
+        {
+            // Ref it...
+            buffer->GetKey()->RefObject();
+
+            // Get the filename from it and add that file if necessary
+            plFileName filename = buffer->GetFileName();
+            if (filename.IsValid())
+            {
+                uint16_t flags = 0;
+
+                if (filename.GetFileExt().compare_i("wav") != 0)
+                {
+                    if (buffer->HasFlag(plSoundBuffer::kOnlyLeftChannel) ||
+                        buffer->HasFlag(plSoundBuffer::kOnlyRightChannel))
+                        hsSetBits(flags, plManifestFile::kSndFlagCacheSplit);
+                    else if (buffer->HasFlag(plSoundBuffer::kStreamCompressed))
+                        hsSetBits(flags, plManifestFile::kSndFlagStreamCompressed);
+                    else
+                        hsSetBits(flags, plManifestFile::kSndFlagCacheStereo);
+                }
+
+                std::pair<ST::string, uint16_t> pair = std::make_pair(filename.AsString(), flags);
+                sfxArray.insert(pair);
+            }
+
+            // Unref the object so it goes away
+            buffer->GetKey()->UnRefObject();
+        }
+    }
+
+    soundKeys.clear();
+
+    plIndirectUnloadIterator iter;
+    rm->IterateAllPages(&iter);
+
+    return true;
+}
+
+
+void DecompressSounds(SoundSet& sounds, bool overwrite, OutputStyle verbosity)
+{
+    uint32_t total = sounds.size();
+    uint32_t curr  = 0;
+
+    if (verbosity == kVerbose)
+    {
+        printf("There are %d sounds\n\n", total);
+    }
+
+    for (auto it = sounds.begin(); it != sounds.end(); ++it)
```

Comment:
We're using C++17 now, so this could be:
`for (const auto& [filename, flags] : sounds)`